### PR TITLE
KCL stdlib reduce function

### PR DESCRIPTION
**Pull request by adamchalmers**
Adds an `arrayReduce` function to KCL stdlib. Right now, it can only reduce SketchGroup values because my implementation of higher-order KCL functions sucks. But we will generalize it in the future to be able to reduce any type.

This simplifies sketching polygons, e.g.

```
fn decagon = (radius) => {
  let step = (1/10) * tau()
  let sketch = startSketchAt([
    (cos(0) * radius), 
    (sin(0) * radius),
  ])
  return arrayReduce([1..10], sketch, (i, sg) => {
      let x = cos(step * i) * radius
      let y = sin(step * i) * radius
      return lineTo([x, y], sg)
  })
}
```

Part of #3842

### CODE REVIEW

**Comment by jessfraz** [comment]
is there another term we can call this, i know jordan was saying `reduce` is a term no ME would intuitively get

File: src/wasm-lib/kcl/src/std/array.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+/// ```
+#[stdlib {
+    name = "arrayReduce",
+}]
+async fn inner_array_reduce<'a>(
+    array: Vec<u64>,
+    start: SketchGroup,
+    reduce_fn: FunctionParam<'a>,
+    args: &'a Args,
+) -> Result<SketchGroup, KclError> {
```
Comment (jtran) [inline src/wasm-lib/kcl/src/std/array.rs:50]: Does the return value need to be a SketchGroup? Can it be a KclValue?

**Comment by adamchalmers** [comment]
> is there another term we can call this, i know jordan was saying `reduce` is a term no ME would intuitively get

`applyTo` because you're applying a function repeatedly? Defined as `applyTo(target, forEach, func)` and used like `applyTo(sketch, [0..10], myFunc)`.

File: src/wasm-lib/kcl/src/std/array.rs (hunk 1 above)
Comment (adamchalmers) [reply src/wasm-lib/kcl/src/std/array.rs:50]: Right now, yes it does, seem PR description :( hopefully we can relax that.

**Comment by jessfraz** [comment]
what about `forEach` anyways im shit at naming things but worth getting @jgomez720 @JordanNoone input

**Comment by jtran** [comment]
> is there another term we can call this, i know jordan was saying `reduce` is a term no ME would intuitively get

The conclusion we came to is that we can make something more user-friendly for MEs, but we'd do that by building on top of lower-level primitives like `reduce`. Only experts should ever need to interact with it. For-loops will essentially compile down to this. It's arguably not for MEs to use. So renaming it is counterproductive, in my opinion.

**Comment by jessfraz** [comment]
ah okay yeah if you give jordan a for loop he will be happy, but if he doesnt have a for loop yet and he tries to use this just be ready :)

File: src/wasm-lib/kcl/src/std/array.rs (hunk 1 above)
Comment (jtran) [reply src/wasm-lib/kcl/src/std/array.rs:50]: Ah, sorry for not reading.

**Review by jtran — APPROVED** [review]

**Comment by jgomez720** [comment]
> is there another term we can call this, i know jordan was saying reduce is a term no ME would intuitively get 

correct I don't know what `reduce` means. But I'm very happy with for loops

**Comment by lf94** [comment]
When the `for` loop "sugar" is introduced, we're going to end up having two ways to do the same thing?

**Pull request merged**